### PR TITLE
Add support for Adtran OS

### PR DESCRIPTION
**Pull request by BSpendlove**
Tested against 2 Adtran TA5000 series, Adtran OS is a generic operating system deploy across all of their devices however I have more TA series that I can test on if need be.

```
Adtran TA5006 and TA5000
Version: 10.3.1.1

$ py.test -v tests/test_netmiko_show.py --test_device adtran_os
================================================ test session starts ================================================
platform linux -- Python 3.6.8, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- /home/brandon/dev/git/netmiko/venv_netmiko/bin/python3
cachedir: .pytest_cache
rootdir: /home/brandon/dev/git/netmiko/netmiko, inifile: setup.cfg
collected 18 items                                                                                                  

tests/test_netmiko_show.py::test_disable_paging PASSED                                                        [  5%]
tests/test_netmiko_show.py::test_ssh_connect PASSED                                                           [ 11%]
tests/test_netmiko_show.py::test_ssh_connect_cm PASSED                                                        [ 16%]
tests/test_netmiko_show.py::test_send_command_timing PASSED                                                   [ 22%]
tests/test_netmiko_show.py::test_send_command PASSED                                                          [ 27%]
tests/test_netmiko_show.py::test_cmd_verify_decorator PASSED                                                  [ 33%]
tests/test_netmiko_show.py::test_send_command_global_cmd_verify PASSED                                        [ 38%]
tests/test_netmiko_show.py::test_send_command_juniper SKIPPED                                                 [ 44%]
tests/test_netmiko_show.py::test_send_command_textfsm SKIPPED                                                 [ 50%]
tests/test_netmiko_show.py::test_send_command_genie SKIPPED                                                   [ 55%]
tests/test_netmiko_show.py::test_base_prompt PASSED                                                           [ 61%]
tests/test_netmiko_show.py::test_strip_prompt PASSED                                                          [ 66%]
tests/test_netmiko_show.py::test_strip_command PASSED                                                         [ 72%]
tests/test_netmiko_show.py::test_normalize_linefeeds PASSED                                                   [ 77%]
tests/test_netmiko_show.py::test_clear_buffer PASSED                                                          [ 83%]
tests/test_netmiko_show.py::test_enable_mode PASSED                                                           [ 88%]
tests/test_netmiko_show.py::test_disconnect PASSED                                                            [ 94%]
tests/test_netmiko_show.py::test_disconnect_no_enable PASSED                                                  [100%]

============================================== short test summary info ==============================================
SKIPPED [1] /home/brandon/dev/git/netmiko/netmiko/tests/test_netmiko_show.py:123: <Skipped instance>
SKIPPED [1] /home/brandon/dev/git/netmiko/netmiko/tests/test_netmiko_show.py:144: TextFSM/ntc-templates not supported on this platform
SKIPPED [1] /home/brandon/dev/git/netmiko/netmiko/tests/test_netmiko_show.py:169: Genie not supported on this platform
===================================== 15 passed, 3 skipped in 104.83s (0:01:44) =====================================

```

It's pretty much 100% the same as Cisco CLI so I've reused the CiscoBaseConnection class

### CODE REVIEW

**Comment by ktbyers** [comment]
@BSpendlove Did you test the configuration operations as well?

py.test -v test_netmiko_config.py --test_device adtran_os

File: netmiko/adtran/adtran.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        """
+        code_cursor_left = chr(27) + r"\[\d+D"
+        output = string_buffer
+        pattern = rf" {code_cursor_left}"
+        output = re.sub(pattern, "", output)
+
+        log.debug("Stripping ANSI escape codes")
+        log.debug(f"new_output = {output}")
+        log.debug(f"repr = {repr(output)}")
+        return super().strip_ansi_escape_codes(output)
```
Comment (ktbyers) [inline netmiko/adtran/adtran.py]: Does Adtran have this behavior also. This was a Huawei oddity we observed recently.

Comment (BSpendlove) [reply netmiko/adtran/adtran.py]: My apologies, I thought it was the exact same issue but it's not exactly the same.

When you write long commands, the prompt will backspace so this:
```
ol03.pag-lab.lon(config-shaper my-test-shaper 1/1)#per interface muxponder-highspeed 1/1/1
```
becomes like this
```
ol03.pag-lab.lon(config-shaper my-test-shaper 1/1)#$face muxponder-highspeed 1/1/1
```

Debug log:
```
DEBUG:netmiko:In strip_ansi_escape_codes
DEBUG:netmiko:repr = 'per interface muxponder-\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08$er interface muxponder-h \x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08$r interface muxponder-hi  \x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08$ inte'
DEBUG:netmiko:Stripping ANSI escape codes
```

**Comment by BSpendlove** [comment]
Regarding the `py.test -v test_netmiko_config.py --test_device adtran_os` please see the results below:
```
$ py.test -v test_netmiko_config.py --test_device adtran_os
================================================ test session starts ================================================
platform linux -- Python 3.6.8, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- /home/brandon/dev/git/netmiko/venv_netmiko/bin/python3
cachedir: .pytest_cache
rootdir: /home/brandon/dev/git/netmiko/netmiko, inifile: setup.cfg
collected 8 items                                                                                                   

test_netmiko_config.py::test_ssh_connect PASSED                                                               [ 12%]
test_netmiko_config.py::test_enable_mode PASSED                                                               [ 25%]
test_netmiko_config.py::test_config_mode PASSED                                                               [ 37%]
test_netmiko_config.py::test_exit_config_mode PASSED                                                          [ 50%]
test_netmiko_config.py::test_config_set PASSED                                                                [ 62%]
test_netmiko_config.py::test_config_set_longcommand PASSED                                                    [ 75%]
test_netmiko_config.py::test_config_from_file PASSED                                                          [ 87%]
test_netmiko_config.py::test_disconnect PASSED                                                                [100%]

=========================================== 8 passed in 102.17s (0:01:42) ===========================================
```

File: netmiko/adtran/adtran.py (hunk 1 above)
Comment (ktbyers) [reply netmiko/adtran/adtran.py]: @BSpendlove Hmmm...that really sounds like an issue/behavior that setting the terminal width should fix. How did you set and test the terminal width?

There is this method you can call in Netmiko

```
strip_backspaces()
```

Here is an example of calling it:

`output = self.strip_backspaces(output)`

Comment (BSpendlove) [reply netmiko/adtran/adtran.py]: Yeah I would of hoped setting the terminal width would fix it, I set it to the max and simply tried sending creating a shaper. I've got a bit further, the commands are actually sent now but the returns output is still a bit shady...

```
def _read_channel_expect(self, *args, **kwargs):
        output = super()._read_channel_expect()

        #Strip backspaces for long commands
        output = self.strip_backspaces(output)
        return output
```

```
DEBUG:netmiko:shaper my-test-shaper 1/1
adtran(config-shaper my-test-shaper 1/1)#per interface muxponder-$er interface muxponder-h $r interface muxponder-hi  $ interface muxponder-hig  $interface muxponder-high  $nterface muxponder-highs  $terface muxponder-highsp  $erface muxponder-highspe  $rface muxponder-highspee  $face muxponder-highspeed  $ace muxponder-highspeed   $ce muxponder-highspeed 1  $e muxponder-highspeed 1/  $ muxponder-highspeed 1/1  $muxponder-highspeed 1/1/  $uxponder-highspeed 1/1/1  
adtran(config-shaper my-test-shaper 1/1)#end
```

I'll get their eventually 😸

Comment (BSpendlove) [reply netmiko/adtran/adtran.py]: @ktbyers I had set the terminal width to the max (132) and created an EVC map on the adtran device, however I found a command that hits the max length and runs into the issue that I found. I've been able to use the `strip_backspaces()` and get rid of the `\x08` but now just figuring out why I'm hitting another issue 👍

**Comment by BSpendlove** [comment]
Sorry I haven't had time to take a look at this still, I've been extremely busy with work, feel free to close this or leave it open, I probably won't be spending much time on it until next month/june

**Comment by BSpendlove** [comment]
If anyone is available to take a look at this issue I am running into, I would gladly appreciate it...

Adtran doesn't have the ability to disable terminal width (like cisco's no terminal width command) so I have to find a solution to properly read the correct prompt. This obviously works eg. in `**send_command**` when I set `**cmd_verify=False**` so that the `**read_until_pattern**` function is skipped, but obviously that seems rather silly to do that...

**Comment by ktbyers** [comment]
One option if there is no `terminal width` capability is it set `global_cmd_verify=False` similar to what was done here:

https://github.com/ktbyers/netmiko/blob/develop/netmiko/cisco/cisco_asa_ssh.py#L28

I don't have a better solution than that if there is no way on the network device to specify the terminal width.

**Comment by ktbyers** [comment]
One other option might be to try to control the width here:

https://github.com/ktbyers/netmiko/blob/v3.2.0/netmiko/base_connection.py#L951

You could see one if this behavior is supported by Adtran and what the maximum value it allows.


Regards, Kirk

**Comment by ktbyers** [comment]
Updated in a newer PR

**Pull request closed without merging**